File: bukkit/libs/guilib/src/main/java/dialight/guilib/gui/Gui.java
package dialight.guilib.gui;

import dialight.guilib.Viewable;
import dialight.guilib.view.View;
import org.bukkit.entity.Player;

// 1 data flow <-> * view
// 1 view <-> * players
public abstract class Gui extends Viewable {

    abstract public View createView(Player player);

}

File: src/main/guilib/dialight/guilib/Gui.kt
package dialight.guilib

import org.spongepowered.api.entity.living.player.Player
import org.spongepowered.api.item.inventory.Inventory

interface Gui {

    fun ownerOf(
        player: Player,
        inv: Inventory
    ): Boolean

    fun getView(player: Player): View
    fun destroyFor(player: Player)

    fun getViewOf(player: Player): View
    fun onCloseView(player: Player): Boolean

}